src/test/java/cloud/eppo/BaseEppoClientTest.java

```diff
@@ -317,6 +343,42 @@ public void testAssignmentEventCorrectlyCreated() {
     assertEquals(expectedMeta, capturedAssignment.getMetaData());
   }
 
+  @Test
+  public void testAssignmentNotDedupedWithoutCache() {
+    initClient();
+
+    Attributes subjectAttributes = new Attributes();
+    subjectAttributes.put("age", EppoValue.valueOf(30));
+    subjectAttributes.put("employer", EppoValue.valueOf("Eppo"));
+
+    // Get the assignment twice
+    eppoClient.getDoubleAssignment("numeric_flag", "alice", subjectAttributes, 0.0);
+    eppoClient.getDoubleAssignment("numeric_flag", "alice", subjectAttributes, 0.0);
+
+    ArgumentCaptor<Assignment> assignmentLogCaptor = ArgumentCaptor.forClass(Assignment.class);
+
+    // `logAssignment` should be called twice;
+    verify(mockAssignmentLogger, times(2)).logAssignment(assignmentLogCaptor.capture());
+  }
+
+  @Test
+  public void testAssignmentEventCorrectlyDeduped() {
+    initClientWithAssignmentCache(new MapAssignmentCache());
+
+    Attributes subjectAttributes = new Attributes();
+    subjectAttributes.put("age", EppoValue.valueOf(30));
+    subjectAttributes.put("employer", EppoValue.valueOf("Eppo"));
+
+    // Get the assignment twice
+    eppoClient.getDoubleAssignment("numeric_flag", "alice", subjectAttributes, 0.0);
+    eppoClient.getDoubleAssignment("numeric_flag", "alice", subjectAttributes, 0.0);
+
+    ArgumentCaptor<Assignment> assignmentLogCaptor = ArgumentCaptor.forClass(Assignment.class);
+
+    // `logAssignment` should be called only once.
+    verify(mockAssignmentLogger, times(1)).logAssignment(assignmentLogCaptor.capture());
```

Comment:
I'd extend this test to then assign a different number, and then the original. Because of the flip-flop, each assignment should be logged as new.

This capture a nuance in that the _key_ should be subject key and flag key, but the value should include the variation. That way new variations evict old ones.

Reply:
added. thanks for the precision.